### PR TITLE
spanner: refactor pooling

### LINKED ISSUE #2113: Spanner queries run from inside transaction callback via non-txn object silently fail
If I'm inside a transaction callback (`runTransaction()`) and I try to execute some query on the database *outside* the transaction, things silently fail.

They should probably either say "you can't do that!" or ... work. Here's a snippet that shows what's happening...

```js
var gcloud = require('google-cloud')({ projectId: 'jjg-cloud-research' });
var spanner = gcloud.spanner();
var database = spanner.instance('test-instance').database('test-database');

database.runTransaction().then(function (data) {
  var txn = data[0];

  var query = 'SELECT * FROM Singers';

  // This works just fine.
  txn.run(query).then(function (results) {
    console.log('txn.run() row count:', results[0].length);
  }).catch(console.log);

  // This should either work or throw an error...
  database.run(query).then(function (results) {
    console.log('database.run() row count:', results[0].length);
  }).catch(console.log);

  // This works just fine.
  var testdb = spanner.instance('test-instance').database('test-database');
  testdb.run(query).then(function (results) {
    console.log('testdb.run() row count:', results[0].length);
  }).catch(console.log);

});
```

I'm not sure what's going on, but the `database.run()` callback never runs:

```
$ node test.js 
txn.run() row count: 3
testdb.run() row count: 3
=
```

Which is super duper weird. Can we either make it throw an error saying "you can't do that outside of a transaction" or make it work as expected? I'd prefer it just work rather than throw an error, but if there's a reason it can't, then I guess we'll survive...

/cc @callmehiphop

### PR DESCRIPTION
**Pull request by callmehiphop**
Closes #2113

#### TODO

- [x] Docs
- [x] Tests

### CODE REVIEW

File: packages/spanner/src/session-pool.js [hunk 1]
```diff
@@ -45,6 +45,8 @@ var util = require('util');
  *     at any given time. (Default: `0`)
  * @param {number} options.keepAlive - How often to ping idle sessions, in
  *     minutes. Must be less than 1 hour.
+ * @param {number} options.acquireTimeout - Time before giving up trying to
```
Comment (stephenplusplus) [inline packages/spanner/src/session-pool.js]: This should either be `acquireTimeoutMs` or just mention we expect this in milliseconds.

**Comment by callmehiphop** [comment]
@stephenplusplus I've updated the unit tests + docs if you could be so kind as to PTAL!

File: packages/spanner/test/session-pool.js [hunk 2]
```diff
@@ -30,7 +30,7 @@ var fakeGenericPool = {
 };
 var originalFakeGenericPool = extend({}, fakeGenericPool);
 
-describe('SessionPool', function() {
+describe.only('SessionPool', function() {
```
Comment (lukesneeringer) [inline packages/spanner/test/session-pool.js]: Remove `.only` before merge. You knew that. :-p

File: packages/spanner/src/session-pool.js [hunk 3]
```diff
@@ -35,18 +35,18 @@ var util = require('util');
  *
  * @param {module:spanner/database} database - The DB instance.
  * @param {object=} options - Configuration options.
+ * @param {number} options.acquireTimeout - Time in milliseconds before giving
+ *     up trying to acquire a session.
  * @param {boolean} options.fail - If set to true, an error will be thrown when
  *     there are no available sessions for a request. (Default: `false`)
  * @param {number} options.max - Maximum number of resources to create at any
- *     given time. (Default: `1`)
+ *     given time. (Default: Infinity)
```
Comment (lukesneeringer) [inline packages/spanner/src/session-pool.js]: I see this documentation change but not the corresponding code.

File: packages/spanner/src/session-pool.js [hunk 4]
```diff
@@ -59,6 +61,9 @@ function SessionPool(database, options) {
   this.maxIdle = options.maxIdle || 1;
   this.fail = !!options.fail;
 
+  this.pendingAcquires = [];
```
Comment (stephenplusplus) [inline packages/spanner/src/session-pool.js]: Quick test needed to make sure we initialize this.

File: packages/spanner/src/session-pool.js [hunk 5]
```diff
@@ -35,10 +35,12 @@ var util = require('util');
  *
  * @param {module:spanner/database} database - The DB instance.
  * @param {object=} options - Configuration options.
+ * @param {number} options.acquireTimeout - Time in milliseconds before giving
+ *     up trying to acquire a session.
```
Comment (stephenplusplus) [inline packages/spanner/src/session-pool.js]: Add a mention for the meaning of `0` and that it is the default value.

File: packages/spanner/src/session-pool.js [hunk 6]
```diff
@@ -35,10 +35,12 @@ var util = require('util');
...
  * @param {boolean} options.fail - If set to true, an error will be thrown when
  *     there are no available sessions for a request. (Default: `false`)
  * @param {number} options.max - Maximum number of resources to create at any
- *     given time. (Default: `1`)
+ *     given time. (Default: Infinity)
```
Comment (stephenplusplus) [inline packages/spanner/src/session-pool.js]: This should say `Number.MAX_SAFE_INTEGER`, I believe.

File: packages/spanner/src/session-pool.js [hunk 7]
```diff
@@ -141,7 +146,7 @@ SessionPool.getPoolOptions_ = function(userOptions) {
   var poolOptions = {
     idleTimeoutMillis: 59 * 60000,
     testOnBorrow: true,
-    max: userOptions.max || 1,
+    max: userOptions.max || Number.MAX_SAFE_INTEGER,
     min: userOptions.min || 0,
     numTestsPerRun: Infinity
```
Comment (stephenplusplus) [inline packages/spanner/src/session-pool.js]: The tests are expecting this to be `Number.MAX_SAFE_INTEGER`

File: packages/spanner/test/session-pool.js [hunk 8]
```diff
@@ -1376,23 +1345,143 @@ describe('SessionPool', function() {
...
-          assert.strictEqual(
-            onAvailable(),
-            sessionPool.getNextAvailableSession_(OPTIONS, assert.ifError)
-          );
+    beforeEach(function() {
+      setInterval_ = global.setInterval;
+      global.setInterval = function(cb) {
+        global.setInterval.calledWith_ = arguments;
+        setImmediate(cb);
+        return 1;
```
Comment (stephenplusplus) [inline packages/spanner/test/session-pool.js]: Can you make this something like `SET_INTERVAL_ID` for clarity when used in the later tests?

File: packages/spanner/src/session-pool.js [hunk 9]
```diff
@@ -489,9 +492,58 @@ SessionPool.prototype.getNextAvailableSession_ = function(options, callback) {
...
   });
+
+  if (this.acquireInterval) {
+    return;
+  }
+
+  var self = this;
+  var intervalSpeed = 30000;
+
+  this.acquireInterval = setInterval(checkForSession, intervalSpeed);
```
Comment (stephenplusplus) [inline packages/spanner/src/session-pool.js]: Can you rename to `this.acquireIntervalId`?

File: packages/spanner/test/session-pool.js [hunk 10]
```diff
@@ -1376,23 +1345,143 @@ describe('SessionPool', function() {
...
+      var intervalArgs = global.setInterval.calledWith_;
+
+      assert.strictEqual(sessionPool.acquireInterval, 1);
+      assert.strictEqual(typeof intervalArgs[0], 'function');
+      assert.strictEqual(intervalArgs[1], 30000);
+    });
+
+    it('should call getNextAvailableSession_ if a read free', function(done) {
+      sessionPool.getNextAvailableSession_ = function(callback) {
+        callback();
```
Comment (stephenplusplus) [inline packages/spanner/test/session-pool.js]: Can you leave a comment here and anywhere else `callback === done` that this is the value of `done`, typically `// done()`

File: packages/spanner/test/session-pool.js [hunk 11]
```diff
@@ -1376,23 +1345,143 @@ describe('SessionPool', function() {
...
+        sessionPool.pendingAcquires = [];
+      };
+
+      global.clearInterval = function(handle) {
+        assert.strictEqual(handle, sessionPool.acquireInterval);
+
+        setImmediate(function() {
+          assert.strictEqual(sessionPool.acquireInterval, null);
+
+          global.clearInterval = clearInterval_;
```
Comment (stephenplusplus) [inline packages/spanner/test/session-pool.js]: Should pop this up to the first line after `global.clearInterval` is re-assigned. (or just capture the real value in the `before` and re-assign in the beforeEach and afterEach)

File: packages/spanner/src/session-pool.js [hunk 12]
```diff
@@ -489,9 +492,58 @@ SessionPool.prototype.getNextAvailableSession_ = function(options, callback) {
...
+    if (!self.pendingAcquires.length) {
+      clearInterval(self.acquireInterval);
+      self.acquireInterval = null;
+    } else if (self.acquireTimeout) {
+      var err = new Error('Unable to acquire Session, timeout occurred.');
+      var acquire;
+
+      for (var i = self.pendingAcquires.length - 1; i > -1; i--) {
+        acquire = self.pendingAcquires[i];
+        acquire.timeout -= intervalSpeed;
```
Comment (stephenplusplus) [inline packages/spanner/src/session-pool.js]: I don't believe this is tested.

File: packages/spanner/src/session-pool.js [hunk 13]
```diff
@@ -489,9 +492,58 @@ SessionPool.prototype.getNextAvailableSession_ = function(options, callback) {
...
+
+        if (acquire.timeout < 0) {
+          acquire.callback(err);
+          self.pendingAcquires.splice(i, 1);
```
Comment (stephenplusplus) [inline packages/spanner/src/session-pool.js]: I don't believe this is tested.

File: packages/spanner/src/session-pool.js (hunk 3 above)
Comment (callmehiphop) [inline packages/spanner/src/session-pool.js]: I think the change you're looking for is here - https://github.com/callmehiphop/gcloud-node/blob/4e63fbf477716b45794e8669ba1c0ca6c21ee567/packages/spanner/src/session-pool.js#L149

File: packages/spanner/test/session-pool.js (hunk 2 above)
Comment (callmehiphop) [inline packages/spanner/test/session-pool.js]: It's a sad day when one of these sneaks by, haha.

File: packages/spanner/src/session-pool.js (hunk 5 above)
Comment (callmehiphop) [inline packages/spanner/src/session-pool.js]: Done!

File: packages/spanner/src/session-pool.js (hunk 6 above)
Comment (callmehiphop) [inline packages/spanner/src/session-pool.js]: Right, fixed!

File: packages/spanner/src/session-pool.js (hunk 4 above)
Comment (callmehiphop) [inline packages/spanner/src/session-pool.js]: Done!

File: packages/spanner/src/session-pool.js (hunk 7 above)
Comment (callmehiphop) [inline packages/spanner/src/session-pool.js]: Fixed!

File: packages/spanner/src/session-pool.js (hunk 9 above)
Comment (callmehiphop) [inline packages/spanner/src/session-pool.js]: Sure can, done!

File: packages/spanner/src/session-pool.js (hunk 13 above)
Comment (callmehiphop) [inline packages/spanner/src/session-pool.js]: Fixed

File: packages/spanner/test/session-pool.js (hunk 11 above)
Comment (callmehiphop) [inline packages/spanner/test/session-pool.js]: Sure can, done!

File: packages/spanner/test/session-pool.js (hunk 10 above)
Comment (callmehiphop) [inline packages/spanner/test/session-pool.js]: Sure, done.

File: packages/spanner/test/session-pool.js (hunk 8 above)
Comment (callmehiphop) [inline packages/spanner/test/session-pool.js]: Sure, done!

File: packages/spanner/src/session-pool.js (hunk 12 above)
Comment (callmehiphop) [inline packages/spanner/src/session-pool.js]: Fixed!

**Comment by stephenplusplus** [comment]
I'm not sure what's up with coverage decreasing. From the Coveralls UI, it looks like, while we are missing coverage in some files, they aren't the ones you edited. Is there an explanation for the coverage dropping that I'm not thinking of?

File: packages/spanner/src/database.js [hunk 14]
```diff
@@ -71,10 +71,13 @@ var TransactionRequest = require('./transaction-request.js');
  *
  * @param {string} name - Name of the database.
  * @param {object=} options - Session pool configuration options.
+ * @param {number} options.acquireTimeout - Time in milliseconds before giving
+ *     up trying to acquire a session. If the specified value is `0`, a timeout
+ *     will not occur.
```
Comment (callmehiphop) [inline packages/spanner/src/database.js]: Done!

**Comment by callmehiphop** [comment]
Not sure, might have something to do with bumping our istanbul version - since we don't really have nightly builds setup, I'd guess this could be the first time coverage has been tested for spanner since upgrading?

**Comment by stephenplusplus** [comment]
Good enough for me, I'll merge after CI's happy.

**Comment by lukesneeringer** [comment]
> I'm not sure what's up with coverage decreasing. From the Coveralls UI, it looks like, while we are missing coverage in some files, they aren't the ones you edited. Is there an explanation for the coverage dropping that I'm not thinking of?

Total and covered lines probably decreased linearly.

**Pull request merged**